neqo-transport/src/server.rs

```diff
@@ -646,37 +630,35 @@ impl Server {
                 return Some(d);
             }
         }
-        qtrace!([self], "No packet to send still, run timers");
-        while let Some(c) = self.timers.take_next(now) {
-            if let Some(d) = self.process_connection(&c, None, now) {
+
+        qtrace!([self], "No packet to send still, check callbacks");
+        loop {
+            let connection = self
+                .connections
+                .borrow()
+                .values()
+                .find(|c| c.borrow().callback.map_or(false, |t| t <= now))
+                .cloned()?;
+            connection.borrow_mut().callback = None;
+            if let Some(d) = self.process_connection(&connection, None, now) {
                 return Some(d);
             }
         }
-        None
     }
 
-    fn next_time(&mut self, now: Instant) -> Option<Duration> {
-        if self.waiting.is_empty() {
-            self.timers.next_time().map(|x| x - now)
-        } else {
-            Some(Duration::new(0, 0))
+    pub fn process(&mut self, dgram: Option<&Datagram>, now: Instant) -> Output {
+        if self.callback.map_or(false, |c| c <= now) {
+            self.callback = None;
         }
-    }
 
-    pub fn process(&mut self, dgram: Option<&Datagram>, now: Instant) -> Output {
         dgram
             .and_then(|d| self.process_input(d, now))
             .or_else(|| self.process_next_output(now))
             .map(|d| {
                 qtrace!([self], "Send packet: {:?}", d);
                 Output::Datagram(d)
             })
-            .or_else(|| {
-                self.next_time(now).map(|delay| {
-                    qtrace!([self], "Wait: {:?}", delay);
-                    Output::Callback(delay)
-                })
-            })
+            .or_else(|| self.callback.take().map(|c| Output::Callback(c - now)))
```

Comment:
Again, this would be better as a function rather than relying on direct access to the field.

Reply:
Note that `self` here is `Server` and not `ServerConnectionState`. Is the additional function on `Server` worth the indirection @martinthomson? Not a strong opinion. Happy to add e.g. the below.

``` diff
modified   neqo-transport/src/server.rs
@@ -15,7 +15,7 @@ use std::{
     ops::{Deref, DerefMut},
     path::PathBuf,
     rc::{Rc, Weak},
-    time::Instant,
+    time::{Duration, Instant},
 };
 
 use neqo_common::{
@@ -661,6 +661,10 @@ impl Server {
         }
     }
 
+    fn take_wake_at(&mut self, now: Instant) -> Option<Duration> {
+        self.wake_at.take().map(|c| c - now)
+    }
+
     pub fn process(&mut self, dgram: Option<&Datagram>, now: Instant) -> Output {
         if self.wake_at.map_or(false, |c| c <= now) {
             self.wake_at = None;
@@ -673,7 +677,7 @@ impl Server {
                 qtrace!([self], "Send packet: {:?}", d);
                 Output::Datagram(d)
             })
-            .or_else(|| self.wake_at.take().map(|c| Output::Callback(c - now)))
+            .or_else(|| self.take_wake_at(now).map(|d| Output::Callback(d)))
             .unwrap_or_else(|| {
                 qtrace!([self], "Go dormant");
                 Output::None
```

Reply:
Nah, that's fine.  It's much easier to process now that the variable is more appropriately named.